examples/ota-provider-app/esp32/main/main.cpp

```diff
@@ -0,0 +1,240 @@
+
+#include <lib/support/ErrorStr.h>
+
+#include <app/clusters/ota-provider/ota-provider.h>
+#include <ota-provider-common/BdxOtaSender.h>
+#include <ota-provider-common/OTAProviderExample.h>
+
+using chip::BitFlags;
+using chip::app::Clusters::OTAProviderDelegate;
+using chip::bdx::TransferControlFlags;
+using chip::Callback::Callback;
+using chip::Messaging::ExchangeManager;
+using namespace ::chip;
+using namespace ::chip::System;
+using namespace ::chip::Credentials;
+using namespace ::chip::DeviceManager;
+using namespace ::chip::DeviceLayer;
+
+const char * TAG = "ota-provider-app";
+
+static DeviceCallbacks EchoCallbacks;
+
+// TODO: this should probably be done dynamically
+constexpr chip::EndpointId kOtaProviderEndpoint = 0;
+
+constexpr uint32_t kMaxBdxBlockSize                = 1024;
+constexpr chip::System::Clock::Timeout kBdxTimeout = chip::System::Clock::Seconds16(5 * 60); // Specifications mandates >= 5 minutes
+constexpr chip::System::Clock::Timeout kBdxPollFreq = chip::System::Clock::Milliseconds32(500);
+const char * gOtaFilename                           = "hello-world.bin";
+BdxOtaSender bdxServer;
+
+const esp_partition_t * otaPartition = nullptr;
+uint32_t otaImageLen                 = 0;
+uint32_t otaTransferInProgress       = false;
+
+void OnBlockQuery(void * context, chip::System::PacketBufferHandle & blockBuf, size_t & size, bool & isEof, uint32_t offset);
+void OnTransferComplete(void * context);
+void OnTransferFailed(void * context);
+
+chip::Callback::Callback<OnBdxBlockQuery> mOnBlockQuery(OnBlockQuery, nullptr);
+chip::Callback::Callback<OnBdxTransferComplete> mOnTransferComplete(OnTransferComplete, nullptr);
+chip::Callback::Callback<OnBdxTransferFailed> mOnTransferFailed(OnTransferFailed, nullptr);
+
+void OnBlockQuery(void * context, chip::System::PacketBufferHandle & blockBuf, size_t & size, bool & isEof, uint32_t offset)
+{
+    if (otaTransferInProgress == false)
+    {
+        if (otaPartition == nullptr || otaImageLen == 0)
+        {
+            ESP_LOGE(TAG, "OTA partition not found");
+            return;
+        }
+        otaTransferInProgress = true;
+    }
+
+    if (offset + kMaxBdxBlockSize >= otaImageLen)
+    {
+        size  = otaImageLen - offset;
+        isEof = true;
+    }
+    else
+    {
+        size  = kMaxBdxBlockSize;
+        isEof = false;
+    }
+
+    esp_err_t err = esp_partition_read(otaPartition, offset + sizeof(otaImageLen), blockBuf->Start(), size);
+    if (err != ESP_OK)
+    {
+        ESP_LOGI(TAG, "Failed to read %d bytes from offset %d", size, offset + sizeof(otaImageLen));
+        size  = 0;
+        isEof = false;
+        return;
+    }
+    ESP_LOGI(TAG, "Read %d bytes from offset %d", size, offset + sizeof(otaImageLen));
+}
+
+void OnTransferComplete(void * context)
+{
+    ESP_LOGI(TAG, "OTA Image Transfer Complete");
+    otaTransferInProgress = false;
+}
+
+void OnTransferFailed(void * context)
+{
+    ESP_LOGI(TAG, "OTA Image Transfer Failed");
```

Comment:
With what error? Should it be emitted to the application?

Reply:
Added status code in the callback.